File: scripts/validate_refactoring.py
#!/usr/bin/env python3
"""
Refactoring Validation Script
Tests the refactored code for basic functionality
"""

import os
import sys
import re

print("=== Ingenium Refactoring Validation ===\n")

tests_passed = 0
tests_failed = 0

def test(name, fn):
    """Run a test and track results"""
    global tests_passed, tests_failed
    
    print(f"Testing: {name}... ", end="")
    try:
        fn()
        print("✅ PASS")
        tests_passed += 1
    except AssertionError as e:
        print(f"❌ FAIL: {e}")
        tests_failed += 1
    except Exception as e:
        print(f"❌ ERROR: {e}")
        tests_failed += 1

def read_file(path):
    """Read file content"""
    with open(path, 'r', encoding='utf-8') as f:
        return f.read()

print("Phase 1: File Logging Utility Tests\n")

# Test 1: File logging utility exists
def test1():
    path = "server/[Tools]/_file_logging.lua"
    assert os.path.exists(path), f"File not found: {path}"

test("File logging utility file exists", test1)

# Test 2: Main logging uses shared utility
def test2():
    content = read_file("server/[Tools]/_logging.lua")
    assert "ig.fileLog.Create" in content, "Should use ig.fileLog.Create"
    assert "local logQueue = {}" not in content, "Should not have old local logQueue"
    assert "local isWriting = false" not in content, "Should not have old isWriting flag"

test("Main logging refactored correctly", test2)

# Test 3: Chat logging uses shared utility
def test3():
    content = read_file("server/_chat.lua")
    assert "ig.fileLog.Create" in content, "Should use ig.fileLog.Create"
    assert "io.open" not in content, "Should not use io.open (FiveM incompatible)"
    assert "local chatLogQueue = {}" not in content, "Should not have old chatLogQueue"

test("Chat logging refactored correctly", test3)

print("\nPhase 2: Gamemode Handler Tests\n")

# Test 4: Gamemode handlers use registry pattern
def test4():
    content = read_file("client/[Events]/_gamemode.lua")
    assert "local gameModeHandlers" in content, "Should have gameModeHandlers registry"
    
    # Should NOT have multiple if blocks
    count = content.count('if conf.gamemode == "RP"')
    assert count == 0, f"Should not have old if conf.gamemode == 'RP' blocks, found {count}"
    
    # Should have single unified handlers
    handler_count = content.count('AddEventHandler("Client:EnteredVehicle"')
    assert handler_count == 1, f"Should have exactly 1 Client:EnteredVehicle handler, found {handler_count}"

test("Gamemode handlers refactored correctly", test4)

print("\nPhase 3: Event Documentation Tests\n")

# Test 5: HUD events have documentation
def test5():
    content = read_file("nui/lua/hud.lua")
    assert "NOTE:" in content and "no registered handlers" in content, "Should document orphaned events"
    assert "TODO:" in content, "Should have TODO for future handlers"

test("HUD events documented", test5)

# Test 6: Drop events documented
def test6():
    content = read_file("client/[Drops]/_drop_integration.lua")
    assert "NOTE:" in content, "Should document event pairing"

test("Drop events documented", test6)

print("\nPhase 4: Documentation Tests\n")

# Test 7: New documentation files exist
def test7a():
    assert os.path.exists("Documentation/FILE_LOGGING_REFACTOR.md"), "Documentation not found"

test("File logging documentation exists", test7a)

def test7b():
    assert os.path.exists("Documentation/GAMEMODE_HANDLER_REFACTOR.md"), "Documentation not found"

test("Gamemode handler documentation exists", test7b)

def test7c():
    assert os.path.exists("Documentation/EVENT_REGISTRATION_ANALYSIS.md"), "Documentation not found"

test("Event analysis documentation exists", test7c)

# Test 8: Main README updated
def test8():
    content = read_file("Documentation/README.md")
    assert "FILE_LOGGING_REFACTOR" in content, "Should link to file logging docs"
    assert "GAMEMODE_HANDLER_REFACTOR" in content, "Should link to gamemode docs"
    assert "EVENT_REGISTRATION_ANALYSIS" in content, "Should link to event analysis"

test("Documentation README updated", test8)

print("\nPhase 5: fxmanifest.lua Tests\n")

# Test 9: fxmanifest includes new file
def test9():
    content = read_file("fxmanifest.lua")
    assert "server/[Tools]/_file_logging.lua" in content, "Should include server/[Tools]/_file_logging.lua"

test("fxmanifest includes file logging utility", test9)

print("\n=== Validation Summary ===")
print(f"✅ Tests Passed: {tests_passed}")
print(f"❌ Tests Failed: {tests_failed}")
print(f"Total: {tests_passed + tests_failed}")

if tests_failed > 0:
    sys.exit(1)
else:
    print("\n🎉 All validation tests passed!")
    sys.exit(0)

File: scripts/validate_refactoring.lua
#!/usr/bin/env lua

-- ====================================================================================--
-- Refactoring Validation Script
-- Tests the refactored code for basic functionality
-- ====================================================================================--

print("=== Ingenium Refactoring Validation ===\n")

local tests_passed = 0
local tests_failed = 0

-- Helper function to test
local function test(name, fn)
    io.write(string.format("Testing: %s... ", name))
    local success, err = pcall(fn)
    if success then
        print("✅ PASS")
        tests_passed = tests_passed + 1
    else
        print("❌ FAIL: " .. tostring(err))
        tests_failed = tests_failed + 1
    end
end

-- Mock ig namespace
_G.ig = _G.ig or {}
_G.ig.fileLog = _G.ig.fileLog or {}
_G.conf = _G.conf or {}

print("Phase 1: File Logging Utility Tests\n")

-- Test 1: File logging utility exists
test("File logging utility file exists", function()
    local f = io.open("server/[Tools]/_file_logging.lua", "r")
    assert(f, "File not found")
    f:close()
end)

-- Test 2: Main logging uses shared utility
test("Main logging refactored correctly", function()
    local f = io.open("server/[Tools]/_logging.lua", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should use ig.fileLog.Create
    assert(content:match("ig%.fileLog%.Create"), "Should use ig.fileLog.Create")
    
    -- Should NOT have old queue code
    assert(not content:match("local logQueue = {}"), "Should not have old local logQueue")
    assert(not content:match("local isWriting = false"), "Should not have old isWriting flag")
end)

-- Test 3: Chat logging uses shared utility
test("Chat logging refactored correctly", function()
    local f = io.open("server/_chat.lua", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should use ig.fileLog.Create
    assert(content:match("ig%.fileLog%.Create"), "Should use ig.fileLog.Create")
    
    -- Should NOT use io.open (FiveM incompatible)
    assert(not content:match("io%.open"), "Should not use io.open")
    
    -- Should NOT have old queue code
    assert(not content:match("local chatLogQueue = {}"), "Should not have old chatLogQueue")
end)

print("\nPhase 2: Gamemode Handler Tests\n")

-- Test 4: Gamemode handlers use registry pattern
test("Gamemode handlers refactored correctly", function()
    local f = io.open("client/[Events]/_gamemode.lua", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should have registry table
    assert(content:match("local gameModeHandlers"), "Should have gameModeHandlers registry")
    
    -- Should NOT have multiple if blocks
    local count = 0
    for _ in content:gmatch('if conf%.gamemode == "RP"') do count = count + 1 end
    assert(count == 0, "Should not have old if conf.gamemode == 'RP' blocks")
    
    -- Should have single unified handlers
    local handler_count = 0
    for _ in content:gmatch('AddEventHandler%("Client:EnteredVehicle"') do 
        handler_count = handler_count + 1 
    end
    assert(handler_count == 1, "Should have exactly 1 Client:EnteredVehicle handler, found " .. handler_count)
end)

print("\nPhase 3: Event Documentation Tests\n")

-- Test 5: HUD events have documentation
test("HUD events documented", function()
    local f = io.open("nui/lua/hud.lua", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should have NOTE comments
    assert(content:match("NOTE:.*no registered handlers"), "Should document orphaned events")
    assert(content:match("TODO:"), "Should have TODO for future handlers")
end)

-- Test 6: Drop events documented
test("Drop events documented", function()
    local f = io.open("client/[Drops]/_drop_integration.lua", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should have NOTE comment
    assert(content:match("NOTE:"), "Should document event pairing")
end)

print("\nPhase 4: Documentation Tests\n")

-- Test 7: New documentation files exist
test("File logging documentation exists", function()
    local f = io.open("Documentation/FILE_LOGGING_REFACTOR.md", "r")
    assert(f, "Documentation not found")
    f:close()
end)

test("Gamemode handler documentation exists", function()
    local f = io.open("Documentation/GAMEMODE_HANDLER_REFACTOR.md", "r")
    assert(f, "Documentation not found")
    f:close()
end)

test("Event analysis documentation exists", function()
    local f = io.open("Documentation/EVENT_REGISTRATION_ANALYSIS.md", "r")
    assert(f, "Documentation not found")
    f:close()
end)

-- Test 8: Main README updated
test("Documentation README updated", function()
    local f = io.open("Documentation/README.md", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should link to new docs
    assert(content:match("FILE_LOGGING_REFACTOR"), "Should link to file logging docs")
    assert(content:match("GAMEMODE_HANDLER_REFACTOR"), "Should link to gamemode docs")
    assert(content:match("EVENT_REGISTRATION_ANALYSIS"), "Should link to event analysis")
end)

print("\nPhase 5: fxmanifest.lua Tests\n")

-- Test 9: fxmanifest includes new file
test("fxmanifest includes file logging utility", function()
    local f = io.open("fxmanifest.lua", "r")
    assert(f, "File not found")
    local content = f:read("*all")
    f:close()
    
    -- Should include new file
    assert(content:match("server/%[Tools%]/_file_logging%.lua"), 
        "Should include server/[Tools]/_file_logging.lua")
end)

print("\n=== Validation Summary ===")
print(string.format("✅ Tests Passed: %d", tests_passed))
print(string.format("❌ Tests Failed: %d", tests_failed))
print(string.format("Total: %d", tests_passed + tests_failed))

if tests_failed > 0 then
    os.exit(1)
else
    print("\n🎉 All validation tests passed!")
    os.exit(0)
end
